File: eloq_store.cpp
#include "eloq_store.h"

#include <glog/logging.h>
#include <sys/resource.h>

#include <atomic>
#include <cassert>
#include <cstddef>
#include <filesystem>
#include <memory>
#include <string_view>
#include <utility>
#include <vector>

#include "archive_crond.h"
#include "common.h"
#include "file_gc.h"
#include "object_store.h"
#include "shard.h"
#include "utils.h"

#ifdef ELOQ_MODULE_ENABLED
#include "eloqstore_module.h"
#endif

namespace eloqstore
{

bool EloqStore::ValidateOptions(const KvOptions &opts)
{
    if ((opts.data_page_size & (page_align - 1)) != 0)
    {
        LOG(ERROR) << "Option data_page_size is not page aligned";
        return false;
    }
    if ((opts.coroutine_stack_size & (page_align - 1)) != 0)
    {
        LOG(ERROR) << "Option coroutine_stack_size is not page aligned";
        return false;
    }

    if (opts.overflow_pointers == 0 ||
        opts.overflow_pointers > max_overflow_pointers)
    {
        LOG(ERROR) << "Invalid option overflow_pointers";
        return false;
    }
    if (opts.max_write_batch_pages == 0)
    {
        LOG(ERROR) << "Invalid option max_write_batch_pages";
        return false;
    }

    if (!opts.cloud_store_path.empty())
    {
        if (opts.local_space_limit == 0)
        {
            LOG(ERROR)
                << "Must set local_space_limit when cloud store is enabled ";
            return false;
        }
        if (!opts.data_append_mode)
        {
            LOG(ERROR) << "append write mode should be enabled when cloud "
                          "storage is enabled";
            return false;
        }
        if (opts.fd_limit * opts.data_page_size *
                (1 << opts.pages_per_file_shift) >
            opts.local_space_limit)
        {
            LOG(ERROR) << "fd_limit * data_page_size * (1 << "
                          "pages_per_file_shift) should be smaller than "
                       << "local_space_limit";
            return false;
        }
    }

    if (opts.data_append_mode)
    {
        if (!opts.cloud_store_path.empty() && opts.DataFileSize() > (8 << 20))
        {
            LOG(WARNING) << "smaller file size is recommended in append write "
                            "mode with cloud storage";
        }
    }
    else
    {
        if (opts.DataFileSize() < (512 << 20))
        {
            LOG(WARNING) << "bigger file size is recommended in non-append "
                            "write mode";
        }
    }
    return true;
}

EloqStore::EloqStore(const KvOptions &opts) : options_(opts), stopped_(true)
{
    if (!ValidateOptions(opts))
    {
        LOG(FATAL) << "Invalid KvOptions configuration";
    }
}

EloqStore::~EloqStore()
{
    if (!IsStopped())
    {
        Stop();
    }
}

KvError EloqStore::Start()
{
    if (!IsStopped())
    {
        LOG(ERROR) << "EloqStore started , do not start again";
        return KvError::NoError;
    }
    eloq_store = this;
    // Initialize
    if (!options_.store_path.empty())
    {
        KvError err = InitStoreSpace();
        CHECK_KV_ERR(err);
    }

    // There are files opened at very early stage like stdin/stdout/stderr, glog
    // file, and root directories of data.
    uint32_t shard_fd_limit = 0;
    size_t used_fd = utils::CountUsedFD();
    if (used_fd + num_reserved_fd < options_.fd_limit)
    {
        shard_fd_limit = (options_.fd_limit - used_fd - num_reserved_fd) /
                         options_.num_threads;
    }

    shards_.resize(options_.num_threads);
    for (size_t i = 0; i < options_.num_threads; i++)
    {
        if (shards_[i] == nullptr)
        {
            shards_[i] = std::make_unique<Shard>(this, i, shard_fd_limit);
        }
        KvError err = shards_[i]->Init();
        CHECK_KV_ERR(err);
    }

    // Start threads.
    stopped_.store(false, std::memory_order_relaxed);

    if (options_.data_append_mode)
    {
        // Initialize file garbage collector for both local and cloud modes
        if (file_gc_ == nullptr)
        {
            file_gc_ = std::make_unique<FileGarbageCollector>(&options_);
        }

        // Only start thread pool in local mode
        if (options_.cloud_store_path.empty() && options_.num_gc_threads > 0)
        {
            LOG(INFO) << "local file gc thread pool started";
            file_gc_->StartLocalThreadPool(options_.num_gc_threads);
        }
        else if (!options_.cloud_store_path.empty())
        {
            LOG(INFO) << "file gc initialized for cloud mode";
            if (options_.num_gc_threads > 0)
            {
                LOG(WARNING)
                    << "num_gc_threads=" << options_.num_gc_threads
                    << " is ignored in cloud mode; GC executes via cloud path.";
            }
        }

        if (options_.num_retained_archives > 0 &&
            options_.archive_interval_secs > 0)
        {
            if (archive_crond_ == nullptr)
            {
                archive_crond_ = std::make_unique<ArchiveCrond>(this);
            }
            archive_crond_->Start();
        }
    }

    for (auto &shard : shards_)
    {
        shard->Start();
    }

#ifdef ELOQ_MODULE_ENABLED
    module_ = std::make_unique<EloqStoreModule>(&shards_);
    eloq::register_module(module_.get());
#endif

    LOG(INFO) << "EloqStore is started.";
    return KvError::NoError;
}

KvError EloqStore::InitStoreSpace()
{
    rlimit fd_limit;
    int res = getrlimit(RLIMIT_NOFILE, &fd_limit);
    if (res < 0)
    {
        LOG(ERROR) << "failed to read open file limit: " << strerror(errno);
        return ToKvError(res);
    }
    DLOG(INFO) << "rlimit open file " << fd_limit.rlim_cur << ", hard limit "
               << fd_limit.rlim_max;
    if (fd_limit.rlim_cur < options_.fd_limit)
    {
        LOG(INFO) << "increase open file limit to " << options_.fd_limit;
        fd_limit.rlim_cur = options_.fd_limit;
        if (setrlimit(RLIMIT_NOFILE, &fd_limit) != 0)
        {
            LOG(ERROR) << "failed to increase open file limit: "
                       << strerror(errno);
            return KvError::OpenFileLimit;
        }
    }

    const bool cloud_store = !options_.cloud_store_path.empty();
    for (const fs::path store_path : options_.store_path)
    {
        if (fs::exists(store_path))
        {
            if (!fs::is_directory(store_path))
            {
                LOG(ERROR) << "path " << store_path << " is not directory";
                return KvError::InvalidArgs;
            }
            if (cloud_store && !std::filesystem::is_empty(store_path))
            {
                LOG(ERROR) << store_path
                           << " is not empty in cloud store mode, clear "
                              "the directory";
                return KvError::InvalidArgs;
            }
            for (auto &ent : fs::directory_iterator{store_path})
            {
                if (!ent.is_directory())
                {
                    LOG(ERROR) << ent.path() << " is not directory";
                    return KvError::InvalidArgs;
                }
            }
        }
        else
        {
            fs::create_directories(store_path);
        }
    }

    assert(root_fds_.empty());
    for (const fs::path store_path : options_.store_path)
    {
        res = open(store_path.c_str(), IouringMgr::oflags_dir);
        if (res < 0)
        {
            for (int fd : root_fds_)
            {
                [[maybe_unused]] int r = close(fd);
                assert(r == 0);
            }
            root_fds_.clear();
            return ToKvError(res);
        }
        root_fds_.push_back(res);
    }
    return KvError::NoError;
}

bool EloqStore::ExecAsyn(KvRequest *req)
{
    req->user_data_ = 0;
    req->callback_ = nullptr;
    return SendRequest(req);
}

void EloqStore::ExecSync(KvRequest *req)
{
    req->user_data_ = 0;
    req->callback_ = nullptr;
    if (SendRequest(req))
    {
        req->Wait();
    }
    else
    {
        req->SetDone(KvError::NotRunning);
    }
}

bool EloqStore::SendRequest(KvRequest *req)
{
    if (stopped_.load(std::memory_order_relaxed))
    {
        return false;
    }

    req->err_ = KvError::NoError;
    req->done_.store(false, std::memory_order_relaxed);

    Shard *shard = shards_[req->TableId().ShardIndex(shards_.size())].get();
    return shard->AddKvRequest(req);
}

void EloqStore::Stop()
{
#ifdef ELOQ_MODULE_ENABLED
    eloq::unregister_module(module_.get());
#endif

    if (archive_crond_ != nullptr)
    {
        archive_crond_->Stop();
    }

    stopped_.store(true, std::memory_order_relaxed);
    for (auto &shard : shards_)
    {
        shard->Stop();
    }

    if (file_gc_ != nullptr)
    {
        file_gc_->Stop();
    }

    // Start clear resources after all threads stopped.

    shards_.clear();

    for (int fd : root_fds_)
    {
        [[maybe_unused]] int res = close(fd);
        assert(res == 0);
    }
    root_fds_.clear();
    LOG(INFO) << "EloqStore is stopped.";
}

const KvOptions &EloqStore::Options() const
{
    return options_;
}

bool EloqStore::IsStopped() const
{
    return stopped_.load(std::memory_order_relaxed);
}

void KvRequest::SetTableId(TableIdent tbl_id)
{
    tbl_id_ = std::move(tbl_id);
}

KvError KvRequest::Error() const
{
    return err_;
}

bool KvRequest::RetryableErr() const
{
    return IsRetryableErr(err_);
}

const char *KvRequest::ErrMessage() const
{
    return ErrorString(err_);
}

uint64_t KvRequest::UserData() const
{
    return user_data_;
}

void KvRequest::Wait() const
{
    CHECK(callback_ == nullptr);
    done_.wait(false, std::memory_order_acquire);
}

void ReadRequest::SetArgs(TableIdent tbl_id, const char *key)
{
    assert(key != nullptr);
    SetArgs(std::move(tbl_id), std::string_view(key));
}

void ReadRequest::SetArgs(TableIdent tbl_id, std::string_view key)
{
    SetTableId(std::move(tbl_id));
    key_.emplace<std::string_view>(key);
}

void ReadRequest::SetArgs(TableIdent tbl_id, std::string key)
{
    SetTableId(std::move(tbl_id));
    key_.emplace<std::string>(std::move(key));
}

std::string_view ReadRequest::Key() const
{
    return key_.index() == 0 ? std::get<std::string_view>(key_)
                             : std::get<std::string>(key_);
}

void FloorRequest::SetArgs(TableIdent tbl_id, const char *key)
{
    assert(key != nullptr);
    SetArgs(std::move(tbl_id), std::string_view(key));
}

void FloorRequest::SetArgs(TableIdent tbl_id, std::string_view key)
{
    SetTableId(std::move(tbl_id));
    key_.emplace<std::string_view>(key);
}

void FloorRequest::SetArgs(TableIdent tbl_id, std::string key)
{
    SetTableId(std::move(tbl_id));
    key_.emplace<std::string>(std::move(key));
}

std::string_view FloorRequest::Key() const
{
    return key_.index() == 0 ? std::get<std::string_view>(key_)
                             : std::get<std::string>(key_);
}

void ScanRequest::SetArgs(TableIdent tbl_id,
                          std::string_view begin,
                          std::string_view end,
                          bool begin_inclusive)
{
    SetTableId(std::move(tbl_id));
    begin_key_.emplace<std::string_view>(begin);
    end_key_.emplace<std::string_view>(end);
    begin_inclusive_ = begin_inclusive;
}

void ScanRequest::SetArgs(TableIdent tbl_id,
                          std::string begin,
                          std::string end,
                          bool begin_inclusive)
{
    SetTableId(std::move(tbl_id));
    begin_key_.emplace<std::string>(std::move(begin));
    end_key_.emplace<std::string>(std::move(end));
    begin_inclusive_ = begin_inclusive;
}

void ScanRequest::SetArgs(TableIdent tbl_id,
                          const char *begin,
                          const char *end,
                          bool begin_inclusive)
{
    std::string_view begin_key = begin == nullptr ? std::string_view{} : begin;
    std::string_view end_key = begin == nullptr ? std::string_view{} : end;
    SetArgs(std::move(tbl_id), begin_key, end_key, begin_inclusive);
}

void ScanRequest::SetPagination(size_t entries, size_t size)
{
    page_entries_ = entries != 0 ? entries : SIZE_MAX;
    page_size_ = size != 0 ? size : SIZE_MAX;

    if (page_entries_ != SIZE_MAX)
    {
        entries_.reserve(page_entries_);
    }
}

std::string_view ScanRequest::BeginKey() const
{
    return begin_key_.index() == 0 ? std::get<std::string_view>(begin_key_)
                                   : std::get<std::string>(begin_key_);
}

std::string_view ScanRequest::EndKey() const
{
    return end_key_.index() == 0 ? std::get<std::string_view>(end_key_)
                                 : std::get<std::string>(end_key_);
}

tcb::span<KvEntry> ScanRequest::Entries()
{
    return tcb::span<KvEntry>(entries_.data(), num_entries_);
}

std::pair<size_t, size_t> ScanRequest::ResultSize() const
{
    size_t size = 0;
    for (size_t i = 0; i < num_entries_; i++)
    {
        const KvEntry &entry = entries_[i];
        size += entry.key_.size() + entry.value_.size();
        size += sizeof(entry.timestamp_) + sizeof(entry.expire_ts_);
    }
    return {num_entries_, size};
}

bool ScanRequest::HasRemaining() const
{
    return has_remaining_;
}

void BatchWriteRequest::SetArgs(TableIdent tbl_id,
                                std::vector<WriteDataEntry> &&batch)
{
    SetTableId(std::move(tbl_id));
    batch_ = std::move(batch);
}

void BatchWriteRequest::AddWrite(std::string key,
                                 std::string value,
                                 uint64_t ts,
                                 WriteOp op)
{
    batch_.push_back({std::move(key), std::move(value), ts, op});
}

void TruncateRequest::SetArgs(TableIdent tbl_id, std::string_view position)
{
    SetTableId(std::move(tbl_id));
    position_ = position;
}

const TableIdent &KvRequest::TableId() const
{
    return tbl_id_;
}

bool KvRequest::IsDone() const
{
    return done_.load(std::memory_order_acquire);
}

void KvRequest::SetDone(KvError err)
{
    err_ = err;
    done_.store(true, std::memory_order_release);
    if (callback_)
    {
        // Asynchronous request
        callback_(this);
    }
    else
    {
        // Synchronous request
        done_.notify_one();
    }
}

}  // namespace eloqstore

File: rustdev/src/store/eloq_store.rs
//! Main EloqStore implementation following C++ eloq_store.cpp

use std::sync::Arc;
use std::sync::atomic::{AtomicBool, Ordering};
use std::os::unix::io::IntoRawFd;
use tokio::task::JoinHandle;

use crate::config::KvOptions;
use crate::error::{Error, KvError};
use crate::shard::Shard;
use crate::Result;

use super::request;

/// Main EloqStore structure
pub struct EloqStore {
    /// Configuration options
    options: Arc<KvOptions>,

    /// Shards for parallel processing
    shards: Vec<Arc<Shard>>,

    /// Shard worker handles
    shard_handles: Vec<JoinHandle<()>>,

    /// Stopped flag
    stopped: AtomicBool,

    /// Root file descriptors for data directories
    root_fds: Vec<i32>,

    // Optional components (following C++)
    // file_gc: Option<FileGarbageCollector>,
    // archive_crond: Option<ArchiveCrond>,
    // obj_store: Option<ObjectStore>,
}

impl EloqStore {
    /// Create a new EloqStore instance
    pub fn new(options: KvOptions) -> Result<Self> {
        // Validate options (following C++ constructor)
        if (options.data_page_size & 0xFFF) != 0 {
            return Err(Error::InvalidConfig("data_page_size must be page aligned (4KB)".into()));
        }

        if options.overflow_pointers == 0 || options.overflow_pointers > crate::types::MAX_OVERFLOW_POINTERS {
            return Err(Error::InvalidConfig("Invalid overflow_pointers".into()));
        }

        if options.max_write_batch_pages == 0 {
            return Err(Error::InvalidConfig("Invalid max_write_batch_pages".into()));
        }

        // Cloud storage validation
        if let Some(ref cloud_path) = options.cloud_store_path {
            if !cloud_path.as_os_str().is_empty() {
                if options.num_gc_threads > 0 {
                    return Err(Error::InvalidConfig(
                        "num_gc_threads must be 0 when cloud store is enabled".into()
                    ));
                }
                if options.local_space_limit == 0 {
                    return Err(Error::InvalidConfig(
                        "Must set local_space_limit when cloud store is enabled".into()
                    ));
                }
            }
        }

        let options = Arc::new(options);

        Ok(Self {
            options,
            shards: Vec::new(),
            shard_handles: Vec::new(),
            stopped: AtomicBool::new(true),
            root_fds: Vec::new(),
        })
    }

    /// Start the store
    pub async fn start(&mut self) -> Result<()> {
        if !self.is_stopped() {
            return Ok(());
        }

        // Initialize store space (following C++ InitStoreSpace)
        self.init_store_space()?;

        // TODO: Initialize cloud storage if configured
        // if let Some(ref cloud_path) = self.options.cloud_store_path {
        //     self.obj_store = Some(ObjectStore::new(&self.options));
        // }

        // Calculate per-shard file descriptor limit
        let used_fd = Self::count_used_fd();
        let reserved_fd = 100; // Reserved FDs for system use
        let shard_fd_limit = if used_fd + reserved_fd < self.options.fd_limit as usize {
            (self.options.fd_limit as usize - used_fd - reserved_fd) / self.options.num_threads
        } else {
            0
        };

        // Create shards (following C++ shard creation)
        self.shards.clear();
        for i in 0..self.options.num_threads {
            let mut shard = Shard::new(
                i,
                self.options.clone(),
                shard_fd_limit as u32,
            );
            shard.init().await?;
            self.shards.push(Arc::new(shard));
        }

        // Mark as started
        self.stopped.store(false, Ordering::Relaxed);

        // Start optional components in append mode
        if self.options.data_append_mode {
            // TODO: Start file GC if configured
            // if self.options.num_gc_threads > 0 {
            //     self.file_gc = Some(FileGarbageCollector::new(&self.options));
            //     self.file_gc.as_mut().unwrap().start(self.options.num_gc_threads);
            // }

            // TODO: Start archive cron if configured
            // if self.options.num_retained_archives > 0 && self.options.archive_interval_secs > 0 {
            //     self.archive_crond = Some(ArchiveCrond::new(self));
            //     self.archive_crond.as_mut().unwrap().start();
            // }
        }

        // TODO: Start cloud storage if configured
        // if let Some(ref mut obj_store) = self.obj_store {
        //     obj_store.start();
        // }

        // Start shard workers
        for shard in &self.shards {
            let shard_clone = shard.clone();
            let handle = tokio::spawn(async move {
                shard_clone.run().await;
            });
            self.shard_handles.push(handle);
        }

        // Give shards a moment to start their run loops
        tokio::time::sleep(tokio::time::Duration::from_millis(10)).await;

        tracing::info!("EloqStore started with {} shards", self.shards.len());
        Ok(())
    }

    /// Stop the store
    pub async fn stop(&mut self) -> Result<()> {
        if self.is_stopped() {
            return Ok(());
        }

        // TODO: Stop archive cron if running
        // if let Some(ref mut archive_crond) = self.archive_crond {
        //     archive_crond.stop().await;
        // }

        // Mark as stopped
        self.stopped.store(true, Ordering::Relaxed);

        // Stop all shards
        for shard in &self.shards {
            shard.stop().await;
        }

        // Wait for shard workers to finish
        for handle in self.shard_handles.drain(..) {
            let _ = handle.await;
        }

        // TODO: Stop cloud storage if configured
        // if let Some(ref mut obj_store) = self.obj_store {
        //     obj_store.stop().await;
        // }

        // TODO: Stop file GC if running
        // if let Some(ref mut file_gc) = self.file_gc {
        //     file_gc.stop().await;
        // }

        // Clear resources
        self.shards.clear();

        // Close root file descriptors
        for fd in self.root_fds.drain(..) {
            unsafe {
                libc::close(fd);
            }
        }

        tracing::info!("EloqStore stopped");
        Ok(())
    }

    /// Read a single key-value pair
    pub async fn read(&self, req: crate::api::request::ReadRequest) -> Result<crate::api::response::ReadResponse> {
        // Convert API request to store request (Key and Value are from bytes::Bytes in both)
        let store_req = Arc::new(request::ReadRequest::new(req.table_id.clone(), req.key));

        // Execute synchronously
        self.exec_sync(store_req.as_ref());

        // Ensure we see all writes from the shard thread
        std::sync::atomic::fence(std::sync::atomic::Ordering::Acquire);

        // Convert response
        if let Some(err) = store_req.base.error.lock().as_ref() {
            return Err(Error::from(*err));
        }

        let value = store_req.value.lock().clone();
        Ok(crate::api::response::ReadResponse {
            value,
        })
    }

    /// Write a batch of key-value pairs
    pub async fn batch_write(&self, req: crate::api::request::BatchWriteRequest) -> Result<crate::api::response::BatchWriteResponse> {
        // Convert API request to store request
        let mut entries = Vec::new();
        for entry in req.entries {
            entries.push(crate::types::WriteDataEntry {
                key: entry.key.to_vec(),
                value: entry.value.to_vec(),
                timestamp: entry.timestamp,
                op: match entry.op {
                    crate::api::request::WriteOp::Upsert => crate::types::WriteOp::Upsert,
                    crate::api::request::WriteOp::Delete => crate::types::WriteOp::Delete,
                },
                expire_ts: entry.expire_ts.unwrap_or(0),
            });
        }

        let store_req = request::BatchWriteRequest::new(req.table_id.clone(), entries);

        // Execute synchronously
        self.exec_sync(&store_req);

        // Ensure we see all writes from the shard thread
        std::sync::atomic::fence(std::sync::atomic::Ordering::Acquire);

        // Convert response
        if let Some(err) = store_req.base.error.lock().as_ref() {
            return Err(Error::from(*err));
        }

        Ok(crate::api::response::BatchWriteResponse {
            success: true,
        })
    }

    /// Scan a range of keys
    pub async fn scan(&self, req: crate::api::request::ScanRequest) -> Result<crate::api::response::ScanResponse> {
        // Convert API request to store request
        let end_key = req.end_key.unwrap_or_else(|| {
            // If no end key, use a max key
            crate::types::Key::from(vec![0xFF; 256])
        });

        let mut store_req = request::ScanRequest::new(
            req.table_id.clone(),
            req.start_key,
            end_key,
            true, // begin_inclusive
        );

        if let Some(limit) = req.limit {
            store_req.set_pagination(limit, 4 * 1024 * 1024);
        }

        // TODO: Handle reverse scanning (store/request doesn't support it directly)

        // Execute synchronously
        self.exec_sync(&store_req);

        // Ensure we see all writes from the shard thread
        std::sync::atomic::fence(std::sync::atomic::Ordering::Acquire);

        // Convert response
        if let Some(err) = store_req.base.error.lock().as_ref() {
            return Err(Error::from(*err));
        }

        // Convert entries from KvEntry to (Key, Value) tuples
        // Note: KvEntry has key and value as Vec<u8>, but we need Bytes
        let mut entries = Vec::new();
        for kv_entry in &store_req.entries {
            entries.push((
                crate::types::Key::from(kv_entry.key.clone()),
                crate::types::Value::from(kv_entry.value.clone()),
            ));
        }

        Ok(crate::api::response::ScanResponse {
            entries,
            has_more: store_req.has_more,
        })
    }

    /// Delete a key
    pub async fn delete(&self, req: crate::api::request::DeleteRequest) -> Result<crate::api::response::DeleteResponse> {
        // Create a delete write entry
        let entry = crate::types::WriteDataEntry {
            key: req.key.to_vec(),
            value: Vec::new(),
            timestamp: 0,
            op: crate::types::WriteOp::Delete,
            expire_ts: 0,
        };

        let store_req = request::BatchWriteRequest::new(req.table_id.clone(), vec![entry]);

        // Execute synchronously
        self.exec_sync(&store_req);

        // Ensure we see all writes from the shard thread
        std::sync::atomic::fence(std::sync::atomic::Ordering::Acquire);

        // Convert response
        if let Some(err) = store_req.base.error.lock().as_ref() {
            return Err(Error::from(*err));
        }

        Ok(crate::api::response::DeleteResponse {
            success: true,
        })
    }

    /// Write a single key-value pair (convenience method)
    pub async fn write(&self, req: crate::api::request::WriteRequest) -> Result<crate::api::response::WriteResponse> {
        let batch_req = crate::api::request::BatchWriteRequest {
            table_id: req.table_id,
            entries: vec![crate::api::request::WriteEntry {
                key: req.key,
                value: req.value,
                op: crate::api::request::WriteOp::Upsert,
                timestamp: 0,
                expire_ts: None,
            }],
            sync: true,
            timeout: None,
        };

        let resp = self.batch_write(batch_req).await?;
        Ok(crate::api::response::WriteResponse {
            success: resp.success,
        })
    }

    /// Execute request asynchronously with callback
    pub async fn exec_async<F>(&self, req: &dyn request::KvRequest, user_data: u64, callback: F) -> bool
    where
        F: FnOnce(&dyn request::KvRequest) + Send + 'static
    {
        // TODO: Set user_data and callback on request
        // req.base.user_data = user_data;
        // req.base.callback = Some(Box::new(callback));
        self.send_request(req).await
    }

    /// Execute request asynchronously without callback
    pub async fn exec_async_no_callback(&self, req: &dyn request::KvRequest) -> bool {
        self.send_request(req).await
    }

    /// Execute request synchronously
    pub fn exec_sync(&self, req: &dyn request::KvRequest) {
        tracing::debug!("exec_sync: sending request type {:?}", req.request_type());
        if !futures::executor::block_on(self.send_request(req)) {
            tracing::debug!("exec_sync: send_request returned false, marking as NotRunning");
            req.set_done(Some(KvError::NotRunning));
        } else {
            tracing::debug!("exec_sync: waiting for request completion");
            req.wait();
            tracing::debug!("exec_sync: request completed");
        }
    }

    /// Send request to appropriate shard
    async fn send_request(&self, req: &dyn request::KvRequest) -> bool {
        if self.stopped.load(Ordering::Relaxed) {
            return false;
        }

        // Don't reset state here - this marks request as done prematurely!

        // Route to appropriate shard based on table ID
        let shard_index = req.table_id().shard_index(self.shards.len() as u16) as usize;
        let shard = &self.shards[shard_index];

        shard.add_request(req).await
    }

    /// Initialize store space (following C++ InitStoreSpace)
    fn init_store_space(&mut self) -> Result<()> {
        // Set file descriptor limit
        #[cfg(unix)]
        {
            use libc::{rlimit, getrlimit, setrlimit, RLIMIT_NOFILE};

            let mut fd_limit = rlimit {
                rlim_cur: 0,
                rlim_max: 0,
            };

            unsafe {
                if getrlimit(RLIMIT_NOFILE, &mut fd_limit) < 0 {
                    return Err(Error::Io(std::io::Error::last_os_error()));
                }

                if fd_limit.rlim_cur < self.options.fd_limit {
                    fd_limit.rlim_cur = self.options.fd_limit;
                    if setrlimit(RLIMIT_NOFILE, &fd_limit) != 0 {
                        return Err(Error::from(KvError::OpenFileLimit));
                    }
                }
            }
        }

        // Create data directories and open root FDs
        let cloud_store = self.options.cloud_store_path.is_some();

        for data_dir in &self.options.data_dirs {
            // Create directory if it doesn't exist
            if !data_dir.exists() {
                std::fs::create_dir_all(data_dir)?;
            }

            // Open directory for O_PATH access (following C++)
            #[cfg(unix)]
            {
                
                use std::fs::OpenOptions;

                let fd = OpenOptions::new()
                    .read(true)
                    .open(data_dir)?
                    .into_raw_fd();

                self.root_fds.push(fd);
            }

            // Set permissions if cloud storage is enabled
            if cloud_store {
                #[cfg(unix)]
                {
                    use std::os::unix::fs::PermissionsExt;
                    let mut perms = std::fs::metadata(data_dir)?.permissions();
                    perms.set_mode(0o755);
                    std::fs::set_permissions(data_dir, perms)?;
                }
            }
        }

        // Create archive directory if configured
        if let Some(ref archive_path) = self.options.archive_path {
            if !archive_path.as_os_str().is_empty() && !archive_path.exists() {
                std::fs::create_dir_all(archive_path)?;
            }
        }

        Ok(())
    }

    /// Count used file descriptors (approximation)
    fn count_used_fd() -> usize {
        #[cfg(unix)]
        {
            if let Ok(entries) = std::fs::read_dir("/proc/self/fd") {
                return entries.count();
            }
        }
        // Default estimate
        10
    }

    /// Check if the store is stopped
    pub fn is_stopped(&self) -> bool {
        self.stopped.load(Ordering::Relaxed)
    }

    /// Get options
    pub fn options(&self) -> &KvOptions {
        &self.options
    }
}

impl Drop for EloqStore {
    fn drop(&mut self) {
        if !self.is_stopped() {
            // Use blocking stop in destructor
            let _ = futures::executor::block_on(self.stop());
        }
    }
}

// Note: Request types are re-exported from store/mod.rs